hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDBCheckpointDiffer.java

```diff
@@ -939,6 +978,114 @@ private void populateCompactionDAG(List<String> inputFiles,
 
   }
 
+  /**
+   * This is the task definition which is run periodically by the service
+   * executor at fixed delay.
+   * It looks for snapshots in compaction DAG which are older than the allowed
+   * time to be in compaction DAG and removes them from the DAG.
+   */
+  public void pruneOlderSnapshotsWithCompactionHistory() {
+    String snapshotDir = null;
+    long currentTimeMillis = System.currentTimeMillis();
+
+    while (!snapshots.isEmpty() &&
+        (currentTimeMillis - snapshots.peek().getLeft())
+            > maxAllowedTimeInDag) {
+      snapshotDir = snapshots.poll().getRight();
+    }
+
+    if (snapshotDir != null) {
+      pruneSnapshotFileNodesFromDag(snapshotDir);
+    }
+  }
+
+  /**
+   * Prunes forward and backward DAGs when oldest snapshot with compaction
+   * history gets deleted.
+   */
+  public void pruneSnapshotFileNodesFromDag(String snapshotDir) {
+    Set<String> snapshotSstFiles = readRocksDBLiveFiles(snapshotDir);
```

Comment:
The pruning is driving by the list of sst files retrieved from the snapshot dirs, but the background deletion task may have already deleted some of those files. Could that lead to gaps in the nodes that get pruned depending on when the background deletion task runs?

Reply:
Good point. This is a valid concern. We might have to read compaction log entries then. Either write the list of SSTs into the compaction log snapshot entry, or ask @sadanand48 's pruning thread to leave a note of the SST list before pruning happens and use that.

Reply:
>  leave a note of the SST list before pruning happens and use that.

We've talked a couple of times about keeping track of state in a new rocksdb instance.

I can't remember where we left that, but if it is being created then that would be a good place for this info.